### PR TITLE
Rover integration testing

### PR DESCRIPTION
**Pull request by markonmars**
This PR adds integration tests for rover liquidations, as well as fixing issues that arose from said testing.

Fixes include
- Incorrect vault pricing
- Prevent attempting to swap through non whitelisted routes
- Prevent attempting to swap via non whitelisted coins
- liquidating gamm pool / lp tokens that were used as collateral (*Not in a vault*)

Also added some optimisations to the base executor, where we now fetch via hive in one requests, where as we were making multiple sequential requests which resulted in slower liquidations

### CODE REVIEW

File: liquidator/test/rover/integration/config.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	    unlockingVault: boolean,
+	    unlockedVault: boolean,
+	    coinBigger: boolean, // todo
+	    vaultBigger: boolean, // todo
+    }
+}
+
+export const config : TestConfig = {
+    seed :
+	'elevator august inherit simple buddy giggle zone despair marine rich swim danger blur people hundred faint ladder wet toe strong blade utility trial process',
```
Comment (elfedy) [inline liquidator/test/rover/integration/config.ts:34]: Is this ok for this seed to be public? Is this a test that any user or a CI process can run? How?

**Review by elfedy — COMMENTED** [review]
Some comments/questions:

Smaller PRs would have been nicer to review. Maybe one PR per test.

Apart from the PR being big, there seems to be a linting thing going on which adds some noise. Would be nice to lint the whole thing in one go, then add rules to the README so everyone knows how to set up things so they are linted from the get go and we don't have unnecessary diffs on the PRs.

Some info on how to run the tests or some context on how they were run designed would be nice. There are some open questions that come to mind when reviewing: 

1.  Are these test meant to be run on testnet? 
2. Are they reproducible by anyone? If so, how would they do so? There seems to be some context assumed on these tests for example the master address having funds and also setting oracle prices but could not find the source of that assumed context. Is this context somewhere in here? How would I go about running those tests

File: liquidator/test/rover/integration/config.ts (hunk 1 above)
Comment (markonmars) [reply liquidator/test/rover/integration/config.ts:34]: Yes, this seed is from [here](https://github.com/mars-protocol/red-bank/blob/master/scripts/deploy/osmosis/config.ts#L25). 

You should be able to run these tests yourself against this testnet deployment as is without any setup. You could run this test in a CI process, however if we want to do that I think we should look at setting up a specific deployment for that so it is sandboxed.

File: liquidator/test/rover/integration/config.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+export interface TestConfig {
```
[the remark this answers is not in the document]
Comment (markonmars) [reply liquidator/test/rover/integration/config.ts:1]: > Are these tests run against testnet?

Well the current config points to testnet but my testing was predominantly done locally as it is much faster with sped up block times.

If you want to test locally, there is a fair bit of setup - you need to deploy both redbank and credit manager, whitelist addresses and set the appropriate addresses and endpoints - hence why I provided a config that utilises testnet, as its plug and play for anybody wanting to run these tests. If you pull down the repo and run `yarn test:rover-integration` the tests should run through without any setup, but to configure a local test requires significant knowledge of the protocol(s) and a LocalOsmosis deployment so takes time and effort for a dev to set up. 

We can look at documenting these steps and automating it as much as possible by utilising the deploy scripts in redbank and rover but that was beyond the scope of this work.

> If so how we guarantee we can reproduce them?

Each test uses a new randomly generated credit manager account to set up a position, create the appropriate conditions and then liquidate. They do not depend on any existing state other than redbank having the required assets and the appropriate liquidity. 

Unfortunately we cannot make the entire test atomic as the liquidator account cannot control the prices. I think creating a specific liquidation testing testnet deployment makes sense, however that was beyond the scope of this work at the time due to time constraints.

> do they actually belong here in the bot repo?

This is a good point. Perhaps we could look at having a separate `liquidation-integration-test` repo, as these tests are not specific to just the liquidation bot but they also are testing rover and apollo's vaults.

**Review by elfedy — APPROVED** [review]

**Comment by markonmars** [comment]
Hey @elfedy thanks for the review

Yea apologies for the large PR - I agree it would be nicer to have test by test or at least split up in some way, although much of the code is around setup of the state which each independent test uses. There is also some a `tsconfig` update in there, as well as some changes to how we store and load our config variables that could have been separate PRs . 

I've made some changes to make the tests as agnostic about starting prices as possible and to support less specific configurations. There are further steps we can take, such as detecting and automating deployments, but I think that is beyond the scope of this PR.

**Pull request merged**